File: plugins/OrchidPages/src/main/java/com/eden/orchid/pages/menu/PagesMenuType.java
package com.eden.orchid.pages.menu;

import com.eden.common.util.EdenUtils;
import com.eden.orchid.api.OrchidContext;
import com.eden.orchid.api.options.annotations.Option;
import com.eden.orchid.api.theme.menus.menuItem.OrchidMenuItem;
import com.eden.orchid.api.theme.menus.menuItem.OrchidMenuItemImpl;
import com.eden.orchid.api.theme.pages.OrchidPage;
import org.apache.commons.lang3.StringUtils;

import javax.inject.Inject;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

public class PagesMenuType extends OrchidMenuItem {

    @Option
    public boolean atRoot;

    @Option
    public String section;

    @Option
    public String title;

    @Inject
    public PagesMenuType(OrchidContext context) {
        super(context, "pages", 100);
    }

    @Override
    public List<OrchidMenuItemImpl> getMenuItems() {
        List<OrchidMenuItemImpl> menuItems = new ArrayList<>();

        List<OrchidPage> allPages = context.getInternalIndex().getGeneratorPages("pages");

        List<OrchidPage> pages = (EdenUtils.isEmpty(section))
                ? allPages
                : allPages
                        .stream()
                        .filter(page -> page.getReference().getPath().startsWith(section))
                        .collect(Collectors.toList());

        if(atRoot) {
            for(OrchidPage page : pages) {
                menuItems.add(new OrchidMenuItemImpl(context, page));
            }
        }
        else {
            if(EdenUtils.isEmpty(title)) {
                if(!EdenUtils.isEmpty(section)) {
                    title = StringUtils.capitalize(section);
                }
                else {
                    title = "Pages";
                }
            }
            menuItems.add(new OrchidMenuItemImpl(context, title, pages));
        }

        for (OrchidMenuItemImpl item : menuItems) {
            item.setIndexComparator(menuItemComparator);
        }

        return menuItems;
    }

    Comparator<OrchidMenuItemImpl> menuItemComparator = (OrchidMenuItemImpl o1, OrchidMenuItemImpl o2) -> {
        String o1Title = "";
        String o2Title = "";

        if (o1.getPage() != null) {
            o1Title = o1.getPage().getTitle();
        }
        else if (o1.isHasChildren() && o1.getChildren().size() > 0 && o1.getChildren().get(0) != null && o1.getChildren().get(0).getPage() != null) {
            o1Title = o1.getChildren().get(0).getTitle();
        }

        if (o2.getPage() != null) {
            o2Title = o2.getPage().getTitle();
        }
        else if (o2.isHasChildren() && o2.getChildren().size() > 0 && o2.getChildren().get(0) != null && o2.getChildren().get(0).getPage() != null) {
            o2Title = o2.getChildren().get(0).getTitle();
        }
        return o1Title.compareTo(o2Title);
    };
}

File: plugins/OrchidPages/src/main/kotlin/com/eden/orchid/pages/menu/PagesMenuType.kt
package com.eden.orchid.pages.menu

import com.eden.common.util.EdenUtils
import com.eden.orchid.api.OrchidContext
import com.eden.orchid.api.generators.OrchidGenerator
import com.eden.orchid.api.options.annotations.Description
import com.eden.orchid.api.options.annotations.Option
import com.eden.orchid.api.theme.menus.MenuItem
import com.eden.orchid.api.theme.menus.OrchidMenuFactory
import org.apache.commons.lang3.StringUtils
import javax.inject.Inject

@Description("Static pages, optionally by group.", name = "Static Pages")
class PagesMenuType
@Inject
constructor(
        context: OrchidContext
) : OrchidMenuFactory(context, "pages", 100) {

    @Option
    @Description("Include only pages in a specific page group, otherwise include all pages.")
    lateinit var group: String

    private var menuItemComparator = Comparator { o1: MenuItem, o2: MenuItem ->
        var o1Title = ""
        var o2Title = ""

        if (o1.page != null) {
            o1Title = o1.page!!.title
        }
        else if (o1.isHasChildren && o1.children.size > 0 && o1.children[0].page != null) {
            o1Title = o1.children[0].title
        }

        if (o2.page != null) {
            o2Title = o2.page!!.title
        }
        else if (o2.isHasChildren && o2.children.size > 0 && o2.children[0].page != null) {
            o2Title = o2.children[0].title
        }

        o1Title.compareTo(o2Title)
    }

    override fun getMenuItems(): List<MenuItem> {
        val menuItems = ArrayList<MenuItem>()

        val allPages = context.index.getChildIndex<OrchidGenerator.Model>("pages")!!.allPages

        val pages = if (EdenUtils.isEmpty(group))
            allPages
        else
            allPages.filter { page -> page.reference.path.startsWith(group) }

        if (EdenUtils.isEmpty(submenuTitle)) {
            if (!EdenUtils.isEmpty(group)) {
                submenuTitle = StringUtils.capitalize(group)
            }
            else {
                submenuTitle = "Pages"
            }
        }
        menuItems.addAll(pages.map {
            MenuItem.Builder(context)
                    .page(it)
                    .indexComparator(menuItemComparator)
                    .build()
        })

        menuItems.sortWith(menuItemComparator)

        return menuItems
    }
}

